libs/langchain/langchain/vectorstores/neo4j_vector.py

```diff
@@ -0,0 +1,626 @@
+from __future__ import annotations
+
+import enum
+import logging
+import uuid
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Iterable,
+    List,
+    Optional,
+    Tuple,
+    Type,
+)
+
+from langchain.docstore.document import Document
+from langchain.embeddings.base import Embeddings
+from langchain.vectorstores.base import VectorStore
+
+
+class DistanceStrategy(str, enum.Enum):
+    """Enumerator of the Distance strategies."""
+
+    EUCLIDEAN = "euclidean"
+    COSINE = "cosine"
+
+
+DEFAULT_DISTANCE_STRATEGY = DistanceStrategy.COSINE
+
+
+class Neo4jVector(VectorStore):
+    """`Neo4j` vector index.
+
+    To use, you should have the ``neo4j`` python package installed.
+
+    Args:
+        url: Neo4j connection url
+        username: Neo4j username.
+        password: Neo4j password
+        database: Optionally provide Neo4j database
+                  Defaults to "neo4j"
+        embedding_function: Any embedding function implementing
+            `langchain.embeddings.base.Embeddings` interface.
+        distance_strategy: The distance strategy to use. (default: COSINE)
+        pre_delete_collection: If True, will delete existing data if it exists.
+            (default: False). Useful for testing.
+
+    Example:
+        .. code-block:: python
+
+            from langchain.vectorstores.neo4j_vector import Neo4jVector
+            from langchain.embeddings.openai import OpenAIEmbeddings
+
+            url="bolt://localhost:7687"
+            username="neo4j"
+            password="pleaseletmein"
+            embeddings = OpenAIEmbeddings()
+            vectorestore = Neo4jVector.from_documents(
+                embedding=embeddings,
+                documents=docs,
+                url=url
+                username=username,
+                password=password,
+            )
+
+
+    """
+
+    def __init__(
+        self,
+        username: str,
+        password: str,
+        url: str,
+        embedding_function: Embeddings,
+        database: str = "neo4j",
+        index_name: str = "vector",
+        node_label: str = "Chunk",
+        embedding_node_property: str = "embedding",
+        text_node_property: str = "text",
+        distance_strategy: DistanceStrategy = DEFAULT_DISTANCE_STRATEGY,
+        logger: Optional[logging.Logger] = None,
+        pre_delete_collection: bool = False,
+        relevance_score_fn: Optional[Callable[[float], float]] = None,
+    ) -> None:
+        try:
+            import neo4j
+        except ImportError:
+            raise ValueError(
+                "Could not import neo4j python package. "
+                "Please install it with `pip install neo4j`."
+            )
+
+        self._driver = neo4j.GraphDatabase.driver(url, auth=(username, password))
+        self._database = database
+        self.schema = ""
+        # Verify connection
+        try:
+            self._driver.verify_connectivity()
```

Comment:
not sure we want to do this on instantiation, maybe this is its own method that users are recommended to call themselves? cc @eyurtsev

Reply:
Since most of the time, the users are using  `from_texts` or `from_documents` method to instantiate, I think it makes sense to leave it here

Reply:
Also, directly after this method we fetch relevant information from the database, so if it didnt' fail here, it would fail directly in the next step (if Neo4j was unavailable or wrong authorization, etc...)